testkit/src/checkpoint_db.rs

```diff
@@ -128,66 +151,28 @@ impl<T: Database> CheckpointDbInner<T> {
     }
 
     fn merge(&mut self, patch: Patch) -> StorageResult<()> {
-        if self.backup_stack.is_empty() {
-            self.db.merge(patch)
-        } else {
-            self.merge_with_logging(patch)
-        }
-    }
-
-    fn merge_with_logging(&mut self, patch: Patch) -> StorageResult<()> {
-        // NB: make sure that **both** the db and the journal
-        // are updated atomically.
-        let snapshot = self.db.snapshot();
-
-        let patch_changes = patch.changes();
-        self.db.merge(patch)?;
-
-        let rev_fork = self.db.fork();
-
-        // Reverse a patch to get a backup patch.
-        for (name, changes) in &patch_changes {
-            let mut view = View::new(&rev_fork, name.clone());
-
-            for (key, _) in changes.iter() {
-                match snapshot.get(name, key) {
-                    Some(value) => {
-                        view.put(key, value);
-                    }
-                    None => {
-                        view.remove(key);
-                    }
-                }
-            }
-        }
-
-        self.backup_stack
-            .last_mut()
-            .expect("`merge_with_logging` called before checkpoint has been set")
-            .push(rev_fork.into_patch());
-        Ok(())
+        self.db.merge(patch)
     }
 
     fn checkpoint(&mut self) {
-        self.backup_stack.push(Vec::new())
+        let name: String = Alphanumeric.sample_iter(thread_rng()).take(32).collect();
+        let checkpoint = self
+            .db
+            .create_checkpoint(&*self.temp_dir.path().join(name))
```

Comment:
Why not create fresh temp_dir for each checkpoint?

Reply:
To store one `TempDir` instead of one per checkpoint.

Reply:
TempDir is just a `Option<PathBuf>`, according to the [source code](https://docs.rs/tempfile/3.1.0/src/tempfile/dir.rs.html#194-196) of the tempfile. Here you're implementing part of the `tempfile` functionality yourself by involving additional dependency.